File: src/main/java/org/gecko/actions/DeleteRegionViewModelElementAction.java
package org.gecko.actions;

import org.gecko.exceptions.GeckoException;
import org.gecko.model.Automaton;
import org.gecko.viewmodel.GeckoViewModel;
import org.gecko.viewmodel.PositionableViewModelElement;
import org.gecko.viewmodel.RegionViewModel;

public class DeleteRegionViewModelElementAction extends AbstractPositionableViewModelElementAction {
    private final GeckoViewModel geckoViewModel;
    private final RegionViewModel regionViewModel;
    private final Automaton automaton;

    DeleteRegionViewModelElementAction(
        GeckoViewModel geckoViewModel, RegionViewModel regionViewModel, Automaton automaton) {
        this.geckoViewModel = geckoViewModel;
        this.regionViewModel = regionViewModel;
        this.automaton = automaton;
    }

    @Override
    boolean run() throws GeckoException {
        automaton.removeRegion(regionViewModel.getTarget());
        geckoViewModel.deleteViewModelElement(regionViewModel);
        return true;
    }

    @Override
    Action getUndoAction(ActionFactory actionFactory) {
        return new RestoreRegionViewModelElementAction(geckoViewModel, regionViewModel, automaton);
    }

    @Override
    PositionableViewModelElement<?> getTarget() {
        return regionViewModel;
    }
}

File: src/main/kotlin/org/gecko/actions/DeleteRegionViewModelElementAction.kt
package org.gecko.actions

import org.gecko.exceptions.GeckoException
import org.gecko.viewmodel.*

/**
 * A concrete representation of an [Action] that removes a [Region] from the [GModel]
 * and its target-[org.gecko.model.Region] from the given [Automaton].
 */
class DeleteRegionViewModelElementAction internal constructor(
    val gModel: GModel,
    val Region: Region,
    val automaton: Automaton
) : AbstractPositionableViewModelElementAction() {
    @Throws(GeckoException::class)
    override fun run(): Boolean {
        automaton.removeRegion(Region)
        val states: List<StateViewModel> = ArrayList(Region.statesProperty)

        for (state in states) {
            Region.removeState(state)
        }

        gModel.deleteViewModelElement(Region)
        gModel.currentEditor!!.updateRegions()
        return true
    }

    override fun getUndoAction(actionFactory: ActionFactory): Action {
        return RestoreRegionViewModelElementAction(gModel, Region, automaton)
    }

    override val target: PositionableViewModelElement
        get() = Region
}
